### ISSUE TITLE
[Question] IP address for agents registered with "any"?

### ISSUE THREAD

**Issue by lucanus81** [wazuh/wazuh#5024]
Hello,
I have a question. I really love the possibility to register an agent using "any" IP, however I still need to figure out what's the temporary IP assigned to an agent.
While "grepping" the whole folder "/var/ossec" I was actually surprised to see that Wazuh stores its data into an sqllite3 database. In particular, 
`[root@localhost db]# pwd
/var/ossec/var/db
[root@localhost db]# sqlite3 global.db "select ip,register_ip from agent where id='003';"
172.16.204.220|any`

I found that var/db/global.db can be queried (in particular the table "agent" has exactly what I need). I wanted to know: how reliable is this technique? Can I rely on it, or one day you might decide to change completely this information?

To be honest, the best approach would be to enhance bin/agent_control -l to show this information as well. If you agree, I could even make a pull request for it :)

[**update**]I was just checking the c++ code for agent_control, and it might not be an easy change to add the information about the real ip. When listing all agents you're pretty much relying on AUTH_FILE, namely etc/client.keys (and in there there's nothing but "any")

Anyway, thanks for the great work you're doing.

Luca

**Comment by vikman90** [comment]
Hi @lucanus81,

Thank you for your words! The `ip` and the `register_ip` columns have a different proposal. First of all, the database at _/var/ossec/var/db_ is a "cache" to be used by the API. I don't recommend you access the SQLite databases while the manager is running, because that may interfere with the Wazuh DB daemon or the API.

Let me explain to you the difference between these two fields:

The `register_ip` field is the IP, or the IP range (like `192.168.10.0/24`) which an agent is allowed to connect from. It may be `any`, meaning `0.0.0.0/32` (any IPv4).

This data comes from the file _/var/ossec/etc/client.keys_:
```sh
# cat /var/ossec/etc/client.keys
001 centos8 any 86198756973823f7dec01b564f7d5e886b0f87a57c8c7db1eca9ea18f141ca99
```

On the other hand, `ip` is the current IP address assigned to the agent's main network interface. This data is reported by the agent, and originally stored into the _agent info_ file:

```sh
# cat /var/ossec/queue/agent-info/centos8-any
Linux |centos8 |4.18.0-80.el8.x86_64 |#1 SMP Tue Jun 4 09:19:46 UTC 2019 |x86_64 [CentOS Linux|centos: 8.0] - Wazuh v3.13.0
01b1f86c766dbf69837d599f30c18045 merged.mg
#"_agent_ip":10.0.2.15

#"_manager_hostname":focal
#"_node_name":node01
```

However, the best method to gather this data is through the API:
```json
# curl -u foo:bar -k https://127.0.0.1:55000/agents/001?pretty
{
   "error": 0,
   "data": {
      "group": [
         "default"
      ],
      "node_name": "node01",
      "configSum": "ab73af41699f13fdd81903b5f23d8d00",
      "os": {
         "arch": "x86_64",
         "major": "8",
         "minor": "0",
         "name": "CentOS Linux",
         "platform": "centos",
         "uname": "Linux |centos8 |4.18.0-80.el8.x86_64 |#1 SMP Tue Jun 4 09:19:46 UTC 2019 |x86_64",
         "version": "8.0"
      },
      "name": "centos8",
      "dateAdd": "1970-01-01 00:00:00",
      "status": "Active",
      "ip": "10.0.2.15",
      "lastKeepAlive": "2020-05-13 12:47:55",
      "version": "Wazuh v3.13.0",
      "mergedSum": "8b40b152365d7f8a6d5e92272c0fcf6c",
      "id": "001",
      "manager": "focal",
      "registerIP": "any"
   }
}
```

In this case, `registerIP` is `any`, meaning that the agent connection is not restricted to a specific address. `ip` is "10.0.2.15", the primary agent's IP address.

Hope this helps you.
Best regards

**Comment by lucanus81** [comment]
@vikman90,

thanks! Unfortunately we cannot use the REST API in our case, so I need to find an alternative way. So can I assume that the way you recommend to get the real IP currently assigned to an "any" agent would be to
- use ./agent_control -l to gather the NAME assigned to an agent
- extract from /var/ossec/queue/agent-info/${NAME}-any the "#"_agent_ip" field?

Thanks,
Luca

**Comment by vikman90** [comment]
Hi @lucanus81,

Yes, that method is good. As I told you yesterday, the SQLite 3 database acts as a "cache" for the API: _wazuh-modulesd_ reads _client.keys_, _agent-info_ and other files and updates the _global.db_ database. On the other hand, the API reads that database.

SQLite3 does not support concurrent access to a database, so we enabled WAL mode in the _global.db_ database so that SQLite3 will reject any operation if another process is accessing the database. Both _wazuh-modulesd_ and the API were designed to detect this condition and retry the operation.

In other words, if you query the database, you might receive an error like this:
```sh
sqlite3 /var/ossec/var/db/global.db "select * from agent"
Error: database is locked
```

If you can deal with this error and retry the operation, you're OK to use the database for your convenience.

Best regards.

**Comment by lucanus81** [comment]
@vikman90,

wonderful! Thanks for your confirmation.
Feel free to close this ticket.

Thanks,
Luca
